### ISSUE TITLE
Sync Messages

### ISSUE THREAD

**Issue by andreievg** [openmsupply/mobile#1849]
## Is your feature request related to a problem? Please describe.

Need a way to re-sync extra fields/tables when mobile is upgraded, this happens when server is upgraded and configured prior to mobile app, i.e.
* Example would be for Dispensing or Vaccine Module, when we upgrade mobile to that version, a message is sent to desktop to re-sync related tables etc.. 

## Describe the solution you'd like

For now we only need one way sync of messages (from mobile to desktop)

## Implementation

* Create messages table
* In migration code add a message for every version upgrade that may require resync of certain tables:

message.type = "mobile_upgrade" (this is what desktop side of things checks, can be translated in sync translator if you want mobile schema to be "versionUpgrade")
message.body = "{ toVersion: '${new APP_VERSION_KEY}', fromVersion: '${APP_VERSION_KEY}', syncSiteId: ${SYNC_SITE_ID} }" (note syncSiteId is a number, others are text)

## Describe alternatives you've considered

Only other option is to reinstall the app when it's upgraded, which will trigger full resync (not that desirable)

## Additional context

mSupply issue: https://github.com/sussol/msupply/issues/5038

Priority set to 'high' as we foresee a lot of manual resync if this is not implemented.

**Comment by wlthomson** [comment]
Just putting this out there for consideration.

One other option would be to add a third optional versioning query parameter to the `sync/v3/initial_dump` endpoint specifying the current mobile version? E.g. something like:


```
http://localhost:2048/sync/v3/initial_dump?from_site=2&to_site=1&schemaVersion=10
```

I don't think it would be too tricky to implement server-side, and has the benefit of being consistent with the current mobile sync pattern and possibly less work?

**Comment by joshxg** [comment]
@sworup for help with this one! If we were to implement the same `message` table in desktop in mobile, it seems to be quite a simple - just add `Message` to schema, and `outgoingSyncUtils` to send the message to the server?

**Comment by Chris-Petty** [comment]
@wlthomson Downside there is mobile still needs to be offline first (even if usually an update is pushed over the internet). Extending and using the REST API in this way would need some additional doodads in the sync logic, where as the messages table has more immediate uses without touching sync or API code (like if authorisation came to mobile)

**Comment by andreievg** [comment]
https://github.com/openmsupply/mobile/issues/1849#issuecomment-570998202

@wlthomson kind of a good point, i think the issue is with code reusability, and i think future additions to method that is used in initial_dump (`sync_create_records_for_store`), should involve added versioning methods that can be used in the message processing logic on the server, i.e.:
`sync_create_records_for_store`
```
...
...
initialiseMobileV3.2()
...
...
```

**Comment by wlthomson** [comment]
> [#1849 (comment)](https://github.com/openmsupply/mobile/issues/1849#issuecomment-570998202)
> 
> @wlthomson kind of a good point, i think the issue is with code reusability, and i think future additions to method that is used in initial_dump (`sync_create_records_for_store`), should involve added versioning methods that can be used in the message processing logic on the server, i.e.:
> `sync_create_records_for_store`
> 
> ```
> ...
> ...
> initialiseMobileV3.2(storeID) // Add v3.2 specific 'stuff' to sync queue
> ...
> ...
> ```
> 
> `message_onUpgradeMobileSite`
> 
> ```
> ...
> initialiseMobileV3.2(storeID) // Add v3.2 specific 'stuff' to sync queue
> ...
> ```

Yeah I agree with some methods, if not only because the `sync_create_records_for_X` methods are pretty big. If those methods are extracted out as suggested, as far as I can see, either implementation (messages or API) will maintain a similar amount of modularisation/re-usability 

The pros I can think of for API:

- Consistent sync "pattern".
- No existing mobile implementation (+ additional tables required).
- Generalises existing initialisation logic (`initial_dump` + `queued_records`), so probably a smaller code overhead (at least mobile side).

The main pro I can think of for messages is that an implementation on desktop already exists.

I can see the API-approach being relatively low-cost (add a couple of optional query params, add a case statement which falls through `initialiseMobile...` calls in same way as `startup_update_version`), but if messages is the way to go I'm not too worried.

**Comment by andreievg** [comment]
@wlthomson 

> Consistent sync "pattern"

Not sure what this means, do you mean `would use existing api in an extended way`
Sync messages don't break any existing existing sync patterns/rules btw.

> No existing mobile implementation (+ additional tables required).

There is a bit of implementation, especially when we put in question like:

Should the extended initial_dump api be called before or after all records in current sync queue are synced to server ? 

> Generalises existing initialisation logic (initial_dump + queued_records), so probably a smaller code overhead (at least mobile side).

If we are reusing existing initial + queued_records logic, we need to consider how initialisation phase blocks users from using mobile and modify things to allow of line use after upgrade (if internet is lacking).

> The main pro I can think of for messages is that an implementation on desktop already exists.

I would argue that effort wise, we would get more bang for our time. Another benefit is flexibility and no changes to existing sync ecosystem.

Btw we only need to add a message table, sync translators and maybe a helper method for creating messages

**Comment by wlthomson** [comment]
> @wlthomson
> 
> > Consistent sync "pattern"
> 
> Not sure what this means, do you mean `would use existing api in an extended way`
> Sync messages don't break any existing existing sync patterns/rules btw.
> 
> > No existing mobile implementation (+ additional tables required).
> 
> There is a bit of implementation, especially when we put in question like:
> 
> Should the extended initial_dump api be called before or after all records in current sync queue are synced to server ?
> 
> > Generalises existing initialisation logic (initial_dump + queued_records), so probably a smaller code overhead (at least mobile side).
> 
> If we are reusing existing initial + queued_records logic, we need to consider how initialisation phase blocks users from using mobile and modify things to allow of line use after upgrade (if internet is lacking).
> 
> > The main pro I can think of for messages is that an implementation on desktop already exists.
> 
> I would argue that effort wise, we would get more bang for our time. Another benefit is flexibility and no changes to existing sync ecosystem.
> 
> Btw we only need to add a message table, sync translators and maybe a helper method for creating messages

I think since messages have been chosen, the API option is out the window :). But for sake of posterity, a quick summary below:

- I agree messages could be quicker short term, but I think that the time saved is minimal and the API solution would actually be tidier and less work in the medium to long term.
- I think messages could get messy quick, due to them being "asynchronous" and breaking the client-server pattern (how does mobile know that its message was received or understood? how does mobile know when desktop has processed its message?).
- I think the flexibility messages give us is at the cost of splintering our existing communication architecture/protocols, which is not as elegant and is going to require more documentation.
- I think going down the messages path encourages "footrunner" style code in mobile, which gets messy quickly.
- It seems like version upgrade is just a specific case of initialisation, which we already do via the API, so it feels a bit inelegant to implement basically the same logic in two places.

Having said all that, it's not a hill I'm dying on 😄 . The majority seems to be for messages, so I'm in agreement that we just go forward with that.

**Comment by adrianwb** [comment]
Just a reminder that whatever we do on the server side should also work for upgrading desktop remote sites, not just mobile ones...

**Comment by andreievg** [comment]
@adrianwb https://github.com/sussol/msupply/issues/5068

**Comment by joshxg** [comment]
Closed by #1872
